File: src/test/java/com/blackcowmoo/moomark/auth/controller/PssportControllerTest.java
package com.blackcowmoo.moomark.auth.controller;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.blackcowmoo.moomark.auth.model.dto.PassportResponse;
import com.blackcowmoo.moomark.auth.model.entity.User;
import com.blackcowmoo.moomark.auth.model.oauth2.Token;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.springframework.test.web.servlet.MockMvc;
import org.testcontainers.containers.GenericContainer;
import org.testcontainers.containers.MySQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

@SpringBootTest(
  webEnvironment = WebEnvironment.RANDOM_PORT
)
@AutoConfigureMockMvc
@Testcontainers
@Slf4j
public class PssportControllerTest {

  @Container
  private static final MySQLContainer<?> MY_SQL_CONTAINER = new MySQLContainer<>("mysql:8.0.24")
    .withUsername("root")
    .withPassword("root")
    .withInitScript("initDB.sql");

  @Container
  private static final GenericContainer< ? > MY_REDIS_CONTAINER = new GenericContainer<>("redis:6")
    .withExposedPorts(6379);

  @DynamicPropertySource
  public static void properties(DynamicPropertyRegistry registry) {
    registry.add("spring.datasource.url", MY_SQL_CONTAINER::getJdbcUrl);
    registry.add("spring.datasource.username", MY_SQL_CONTAINER::getUsername);
    registry.add("spring.datasource.password", MY_SQL_CONTAINER::getPassword);
//    registry.add("spring.redis.host", MY_REDIS_CONTAINER::getHost);
//    registry.add("spring.redis.port", Integer.toString(6379));
    System.setProperty("spring.redis.host", MY_REDIS_CONTAINER.getHost());
    System.setProperty("spring.redis.port", MY_REDIS_CONTAINER.getMappedPort(6379).toString());
  }

  @Value("${passport.public-key}")
  private String passportPublicKey;

  @Value("${passport.test.token.expired.user}")
  private String expiredTestPassportUser;

  @Value("${passport.test.token.expired.key}")
  private String expiredTestPassportKey;

  @Autowired
  private MockMvc mvc;

  @Autowired
  private ObjectMapper mapper;

  @Test
  void generatePassport() throws Exception {
    String userId = "1234";
    Token token = mapper
      .readValue(mvc.perform(get("/api/v1/oauth2/google").param("code", "test-" + userId))
        .andExpect(status().isOk())
        .andReturn().getResponse().getContentAsString(), Token.class);

    assertNotNull(token.getToken());

    PassportResponse passport = mapper
      .readValue(mvc.perform(get("/api/v1/passport").header("Authorization", token.getToken()))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), PassportResponse.class);

    User user = mapper
      .readValue(mvc.perform(
          get("/api/v1/passport/verify")
            .header("x-moom-passport-user", passport.getPassport())
            .header("x-moom-passport-key", passport.getKey())
        )
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), User.class);

    assertEquals(user.getId(), userId);
  }

  @Test
  void verifyPassport() throws Exception {
    String userId = "1234";
    Token token = mapper
      .readValue(mvc.perform(get("/api/v1/oauth2/google").param("code", "test-" + userId)).andExpect(status().isOk())
        .andReturn().getResponse().getContentAsString(), Token.class);

    assertNotNull(token.getToken());

    PassportResponse passport = mapper
      .readValue(mvc.perform(get("/api/v1/passport").header("Authorization", token.getToken()))
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), PassportResponse.class);

    User user = mapper
      .readValue(mvc.perform(
          get("/api/v1/user")
            .header("x-moom-passport-user", passport.getPassport())
            .header("x-moom-passport-key", passport.getKey())
        )
        .andExpect(status().isOk()).andReturn().getResponse().getContentAsString(), User.class);

    assertEquals(user.getId(), userId);
  }

  @Test
  void checkPublicKey() throws Exception {
    String publicKey = passportPublicKey;
    String testPublicKey = mvc.perform(get("/api/v1/passport/verify/public"))
      .andExpect(status().isOk()).andReturn().getResponse().getContentAsString();

    assertNotNull(publicKey);
    assertNotEquals(publicKey, "");
    assertEquals(publicKey, testPublicKey);
  }

  @Test
  void expiredPassport() throws Exception {
    String response = mvc.perform(
        get("/api/v1/passport/verify")
          .header("x-moom-passport-user", expiredTestPassportUser)
          .header("x-moom-passport-key", expiredTestPassportKey)
      )
      .andExpect(status().isOk()).andReturn().getResponse().getContentAsString();

    assertEquals(response, "");
  }
}

File: src/test/kotlin/com/blackcowmoo/moomark/auth/controller/PssportControllerTest.kt
package com.blackcowmoo.moomark.auth.controller

import com.blackcowmoo.moomark.auth.model.AuthProvider
import com.blackcowmoo.moomark.auth.model.Role
import com.blackcowmoo.moomark.auth.model.dto.PassportResponse
import com.blackcowmoo.moomark.auth.model.entity.User
import com.blackcowmoo.moomark.auth.model.oauth2.Token
import com.blackcowmoo.moomark.auth.service.PassportService
import com.blackcowmoo.moomark.auth.service.TokenService
import com.blackcowmoo.moomark.auth.service.UserService
import com.fasterxml.jackson.databind.ObjectMapper
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.Mockito.anyString
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.TestPropertySource
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status

@SpringBootTest
@AutoConfigureMockMvc
@ActiveProfiles("test")
@TestPropertySource(locations = ["classpath:application-test.yaml"])
class PssportControllerTest {

  @Value("\${passport.public-key}")
  private lateinit var passportPublicKey: String

  @Value("\${passport.test.token.expired.user}")
  private lateinit var expiredTestPassportUser: String

  @Value("\${passport.test.token.expired.key}")
  private lateinit var expiredTestPassportKey: String

  @Autowired
  private lateinit var mvc: MockMvc

  @Autowired
  private lateinit var mapper: ObjectMapper

  @MockBean
  private lateinit var tokenService: TokenService

  @MockBean
  private lateinit var userService: UserService

  @MockBean
  private lateinit var passportService: PassportService

  @Test
  fun generatePassport() {
    val userId = "1234"
    val token = Token("test-jwt-token", "test-refresh-token")
    val user1 = User(userId, AuthProvider.TEST, "test@test.com", "test", "https://test.com", Role.USER)

    `when`(tokenService.generateToken(userId, AuthProvider.TEST, Role.USER)).thenReturn(token)
    `when`(tokenService.verifyToken(anyString())).thenReturn(true)
    `when`(tokenService.getUid(anyString())).thenReturn(userId)
    `when`(tokenService.getProvider(anyString())).thenReturn(AuthProvider.TEST)
    `when`(userService.getUserById(AuthProvider.TEST, userId)).thenReturn(user1)
    `when`(passportService.generatePassport(any())).thenReturn(null)

    val tokenResult = mapper.readValue(
      mvc.perform(get("/api/v1/oauth2/google").param("code", "test-$userId"))
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      Token::class.java
    )

    assertThat(tokenResult.token).isNotNull()

    val passport = mapper.readValue(
      mvc.perform(get("/api/v1/passport").header("Authorization", tokenResult.token))
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      PassportResponse::class.java
    )

    val user2 = mapper.readValue(
      mvc.perform(
        get("/api/v1/passport/verify")
          .header("x-moom-passport-user", passport.passport)
          .header("x-moom-passport-key", passport.key)
      )
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      User::class.java
    )

    assertThat(user2.id).isEqualTo(userId)
  }

  @Test
  fun verifyPassport() {
    val userId = "1234"
    val token = Token("test-jwt-token", "test-refresh-token")
    val user3 = User(userId, AuthProvider.TEST, "test@test.com", "test", "https://test.com", Role.USER)

    `when`(tokenService.generateToken(userId, AuthProvider.TEST, Role.USER)).thenReturn(token)
    `when`(tokenService.verifyToken(anyString())).thenReturn(true)
    `when`(tokenService.getUid(anyString())).thenReturn(userId)
    `when`(tokenService.getProvider(anyString())).thenReturn(AuthProvider.TEST)
    `when`(userService.getUserById(AuthProvider.TEST, userId)).thenReturn(user3)

    val tokenResult = mapper.readValue(
      mvc.perform(get("/api/v1/oauth2/google").param("code", "test-$userId"))
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      Token::class.java
    )

    assertThat(tokenResult.token).isNotNull()

    val passport = mapper.readValue(
      mvc.perform(get("/api/v1/passport").header("Authorization", token.token))
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      PassportResponse::class.java
    )

    val user4 = mapper.readValue(
      mvc.perform(
        get("/api/v1/user")
          .header("x-moom-passport-user", passport.passport)
          .header("x-moom-passport-key", passport.key)
      )
        .andExpect(status().isOk())
        .andReturn().response.contentAsString,
      User::class.java
    )

    assertThat(user4.id).isEqualTo(userId)
  }

  @Test
  fun checkPublicKey() {
    val publicKey = passportPublicKey
    val testPublicKey = mvc.perform(get("/api/v1/passport/verify/public"))
      .andExpect(status().isOk())
      .andReturn().response.contentAsString

    assertThat(publicKey).isNotNull()
    assertThat(publicKey).isNotEmpty()
    assertThat(publicKey).isEqualTo(testPublicKey)
  }

  @Test
  fun expiredPassport() {
    val response = mvc.perform(
      get("/api/v1/passport/verify")
        .header("x-moom-passport-user", expiredTestPassportUser)
        .header("x-moom-passport-key", expiredTestPassportKey)
    )
      .andExpect(status().isOk())
      .andReturn().response.contentAsString

    assertThat(response).isEmpty()
  }
}
